src/mapnik_image.cpp

```diff
@@ -1421,6 +1427,448 @@ Local<Value> Image::_copySync(_NAN_METHOD_ARGS)
+        {
+            int height_tmp = args[1]->IntegerValue();
+            if (height_tmp <= 0)
+            {
+                NanThrowTypeError("Height must be a integer greater then zero");
+                NanReturnUndefined();
+            }
+            height = static_cast<std::size_t>(height_tmp);
+        }
+        else
+        {
+            NanThrowTypeError("Height must be a number");
+            NanReturnUndefined();
+        }
+    }
+    else
+    {
+        NanThrowTypeError("resize requires a width and height paramter.");
+        NanReturnUndefined();
+    }
+    if (args.Length() >= 4)
+    {
+        if (args[2]->IsObject())
+        {
+            options = args[2]->ToObject();
+        }
+        else
+        {
+            NanThrowTypeError("Expected options object as third argument");
+            NanReturnUndefined();
+        }
+    }
+    
+    if (options->Has(NanNew("scaling_method")))
+    {
+        Local<Value> scaling_val = options->Get(NanNew("scaling_method"));
+        if (scaling_val->IsNumber())
+        {
+            scaling_method = static_cast<mapnik::scaling_method_e>(scaling_val->IntegerValue());
+            if (scaling_method > mapnik::SCALING_BLACKMAN)
+            {
+                NanThrowTypeError("Invalid scaling_method");
+                NanReturnUndefined();
+            }
+        }
+        else
+        {
+            NanThrowTypeError("scaling_method argument must be an integer");
+            NanReturnUndefined();
+        }
+    }
+    
+    if (options->Has(NanNew("filter_factor")))
+    {
+        Local<Value> ff_val = options->Get(NanNew("filter_factor"));
+        if (ff_val->IsNumber())
+        {
+            filter_factor = ff_val->NumberValue();
+        }
+        else
+        {
+            NanThrowTypeError("filter_factor argument must be a number");
+            NanReturnUndefined();
+        }
+    }
+    resize_image_baton_t *closure = new resize_image_baton_t();
+    closure->request.data = closure;
+    closure->im1 = im1;
+    closure->scaling_method = scaling_method;
+    closure->size_x = width;
+    closure->size_y = height;
+    closure->filter_factor = filter_factor;
+    closure->error = false;
+    NanAssignPersistent(closure->cb, callback.As<Function>());
+    uv_queue_work(uv_default_loop(), &closure->request, EIO_Resize, (uv_after_work_cb)EIO_AfterResize);
+    closure->im1->Ref();
+    NanReturnUndefined();
+}
+
+struct resize_visitor
+{
+
+    resize_visitor(mapnik::image_any const& im1, 
+                   mapnik::scaling_method_e scaling_method,
+                   double image_ratio_x,
+                   double image_ratio_y,
+                   double filter_factor) :
+        im1_(im1),
+        scaling_method_(scaling_method),
+        image_ratio_x_(image_ratio_x),
+        image_ratio_y_(image_ratio_y) {}
+
+    template <typename T>
+    void operator()(T & im2) const
+    {
+        bool demultiply = mapnik::premultiply_alpha(im2);
```

Comment:
Let's not mutate the image passed in. Instead expect it is premultiplied: https://github.com/mapnik/node-mapnik/issues/470